spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-yarn/src/main/java/org/springframework/cloud/data/module/deployer/yarn/YarnCloudAppStateMachine.java

```diff
@@ -0,0 +1,437 @@
+
+			if (!StringUtils.hasText(appVersion)) {
+				context.getStateMachine().sendEvent(
+						MessageBuilder.withPayload(Events.ERROR).setHeader(HEADER_ERROR, "appVersion not defined")
+								.build());
+			} else {
+				Collection<CloudAppInfo> appInfos = yarnCloudAppService.getApplications();
+				for (CloudAppInfo appInfo : appInfos) {
+					if (appInfo.getName().equals(appVersion)) {
+						context.getExtendedState().getVariables().put(VAR_APP_VERSION, appVersion);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * {@link Guard} which is used to protect state where application push
+	 * into hdfs would happen. Assumes that if {@code appVersion} variable
+	 * exists, application is installed.
+	 */
+	private class PushAppGuard implements Guard<States, Events> {
+
+		@Override
+		public boolean evaluate(StateContext<States, Events> context) {
+			return !context.getExtendedState().getVariables().containsKey(VAR_APP_VERSION);
+		}
+	}
+
+	/**
+	 * {@link Action} which pushes application version into hdfs found
+	 * from variable {@code appVersion}.
+	 */
+	private class PushAppAction implements Action<States, Events> {
+
+		@Override
+		public void execute(StateContext<States, Events> context) {
+			String appVersion = (String) context.getMessageHeader(HEADER_APP_VERSION);
+			yarnCloudAppService.pushApplication(appVersion);
+		}
+	}
+
+	/**
+	 * {@link Action} which queries {@link YarnCloudAppService} for existing
+	 * running instances.
+	 */
+	private class CheckInstanceAction implements Action<States, Events> {
+
+		@Override
+		public void execute(StateContext<States, Events> context) {
+			Collection<CloudAppInstanceInfo> appInstanceInfos = yarnCloudAppService.getInstances();
+			for (CloudAppInstanceInfo appInstanceInfo : appInstanceInfos) {
+				if (appInstanceInfo.getAddress().contains("http")) {
+					context.getExtendedState().getVariables().put(VAR_APPLICATION_ID, appInstanceInfo.getApplicationId());
+					break;
+				}
+			}
+
+		}
+	}
+
+	/**
+	 * {@link Guard} which protects state {@code STARTINSTANCE} in choice state
+	 * {@code STARTINSTANCECHOICE}.
+	 */
+	private class StartInstanceGuard implements Guard<States, Events> {
+
+		@Override
+		public boolean evaluate(StateContext<States, Events> context) {
+			return !context.getExtendedState().getVariables().containsKey(VAR_APPLICATION_ID);
+		}
+	}
+
+	/**
+	 * {@link Action} which launches new application instance.
+	 */
+	private class StartInstanceAction implements Action<States, Events> {
+
+		@Override
+		public void execute(StateContext<States, Events> context) {
+			String appVersion = (String) context.getMessageHeader(HEADER_APP_VERSION);
+			String applicationId = yarnCloudAppService.submitApplication(appVersion);
+			context.getExtendedState().getVariables().put(VAR_APPLICATION_ID, applicationId);
+
+			for (int i = 0; i < 60; i++) {
+				if (isRunning()) {
+					return;
+				}
+				try {
+					Thread.sleep(1000);
+				} catch (InterruptedException e) {
```

Comment:
Swallowing the `InterruptedException` without resetting the interrupt flag will cause this loop to run to completion. Is this the intended behavior? Would it make sense to reset the flag and throw a `RuntimeException`?